src/JBMultiTerminal.sol

```diff
@@ -526,42 +553,66 @@ contract JBMultiTerminal is JBPermissioned, ERC2771Context, IJBMultiTerminal {
+    /// @param accountingContexts The accounting contexts to add.
+    function addAccountingContextsFor(
+        uint256 projectId,
+        JBAccountingContext[] calldata accountingContexts
+    )
+        external
+        override
+    {
         // Enforce permissions.
         _requirePermissionAllowingOverrideFrom({
             account: PROJECTS.ownerOf(projectId),
             projectId: projectId,
             permissionId: JBPermissionIds.ADD_ACCOUNTING_CONTEXTS,
-            alsoGrantAccessIf: _msgSender() == address(DIRECTORY.controllerOf(projectId))
+            alsoGrantAccessIf: _msgSender() == address(_controllerOf(projectId))
         });
 
+        // Get a reference to the project's current ruleset.
+        JBRuleset memory ruleset = RULESETS.currentOf(projectId);
+
+        // Make sure that if there's a ruleset, it allows adding accounting contexts.
+        if (ruleset.id != 0 && !ruleset.allowAddAccountingContext()) revert ADDING_ACCOUNTING_CONTEXT_NOT_ALLOWED();
+
         // Keep a reference to the number of accounting contexts to add.
-        uint256 numberOfAccountingContexts = tokens.length;
+        uint256 numberOfAccountingContexts = accountingContexts.length;
 
-        // Keep a reference to the token being iterated on.
-        address token;
+        // Keep a reference to the accounting context being iterated on.
+        JBAccountingContext memory accountingContext;
 
         // Start accepting each token.
         for (uint256 i; i < numberOfAccountingContexts; i++) {
             // Set the accounting context being iterated on.
-            token = tokens[i];
+            accountingContext = accountingContexts[i];
 
             // Get a storage reference to the currency accounting context for the token.
-            JBAccountingContext storage accountingContext = _accountingContextForTokenOf[projectId][token];
+            JBAccountingContext storage storedAccountingContext =
+                _accountingContextForTokenOf[projectId][accountingContext.token];
 
             // Make sure the token accounting context isn't already set.
-            if (accountingContext.token != address(0)) revert ACCOUNTING_CONTEXT_ALREADY_SET();
+            if (storedAccountingContext.token != address(0)) revert ACCOUNTING_CONTEXT_ALREADY_SET();
+
+            // Make sure the decimals provided in the accounting context matches expectations if there are expectations.
+            if (
+                (accountingContext.token == JBConstants.NATIVE_TOKEN && accountingContext.decimals != 18)
+                    || (
+                        IERC165(accountingContext.token).supportsInterface(type(IERC20Metadata).interfaceId)
+                            && accountingContext.decimals != IERC20Metadata(accountingContext.token).decimals()
+                    )
+            ) {
+                revert INVALID_ACCOUNTING_CONTEXT_DECIMALS();
```

Comment:
nit: we could add an argument to this error to specify which token was passed incorrectly